[RFC] Add a way to color command-line

ZyX-I opened the pull request:
Based on #5119 as I do not want merge conflicts.

The idea is that we can color Neovim command-line. Ultimate goal is

1. Refactor VimL parsing so it can produce information about colors (for `:`/command prompts).
2. Refactor VimL expression parsing so it can produce informaiton about colors (for `=`/expr prompts).
3. Add a way for the user to specify how his input() prompts should be colored.
4. Add a way for the user to specify how his custom commands should be colored.

This is logical to split that into the following PRs:

1. (This) add basic prompt coloring functionality. Adds a few callbacks, including one which may be used to make pygments color Ex commands and expressions for us. Also add a way for the user to color input() prompts. If some strange problems will not occur this part is going to be easy.
2. Remove callback which colors expressions, move expressions parsing to `src/nvim/eval` and split it from execution, make it produce coloring information. Obsoletes or reuses expression parser from #243. Reuses tests from #243. Resulting code is going to be rather unoptimal: expressions are going to be reparsed each time.
3. (Many PRs) port commands parsing from #243; most likely one-by-one since #243 is not good for highlighting and is rather outdated. Execute commands with C, no translation. Again reparse things each time up until all commands are ported.
4. Actually build AST for functions and execute it without reparsing. Though this has nothing to do with command-line coloring.
5. Only now provide `:command -color=funcname` for coloring user commands.

---

Proposed API for input():

`input(prompt :: String[, text :: String[, completion :: String]])`: no additions.

`input(prompt :: String, options :: Dictionary)`, `options` allow keys text, completion and highlight. No third argument allowed.

ZyX-I committed:
ex_getln: Clean up draw_cmdline a bit

ZyX-I committed:
ex_getln: Do not do arabic shaping unless needed

Should speed up execution without arabic characters a bit, slowing down with 
arabic characters. More necessary, this allows coloring prompt without caring 
about arabic shaping at the first iteration.

ZyX-I committed:
ex_getln: Add basic support for coloring command-line prompt

ZyX-I committed:
functests: Start adding some tests

ZyX-I commented:
PR commits  start with “ex_getln: Clean up draw_cmdline a bit”.

ZyX-I commented:
Example code which will highlight using pygments:

```Python
# color_cmdline.py
def color_cmdline_setup():
    import vim
    from pygments import lex, format
    from pygments.lexers import get_lexer_by_name
    from pygments.formatter import Formatter
    from pygments.token import Token

    try:
        from __builtin__ import unicode
    except ImportError:
        unicode = str

    COLORS = {
        Token:                    'Normal',

        Token.Comment:            'Comment',
        Token.Comment.Preproc:    'PreProc',
        Token.Keyword:            'Keyword',
        Token.Keyword.Type:       'Type',
        Token.Operator:           'Operator',
        Token.Name:               'Identifier',
        Token.Name.Builtin:       'Statement',
        Token.Name.Function:      'Function',
        Token.Name.Constant:      'Number',
        Token.String:             'String',
        Token.Number:             'Number',
        Token.Punctuation:        'Operator',

        Token.Error:              'Error',
    }

    class NvimFormatter(Formatter):
        def __init__(self, **kwargs):
            super(NvimFormatter, self).__init__(**kwargs)
            self.offset = 0
            self.colorscheme = kwargs.get('colorscheme', None) or COLORS

        def format_unencoded(self, tokensource, outfile):
            for ttype, value in tokensource:
                ttype_str = str(ttype).replace(' ', '_')
                value = value.rstrip('\n')
                color = self.colorscheme.get(ttype)
                while color is None:
                    ttype = ttype[:-1]
                    color = self.colorscheme.get(ttype)
                if isinstance(value, unicode):
                    vlen = len(value.encode('utf-8'))
                else:
                    vlen = len(value)
                if vlen:
                    outfile.write('{} {} {} {}\n'.format(
                        self.offset, self.offset + vlen, color, ttype_str))
                self.offset += vlen

    def setup():
        global color_cmdline_highlight
        lexer = get_lexer_by_name('vim')

        def color_cmdline_highlight():
            cmdline = vim.eval('a:cmdline')
            add_tt = False
            if bool(int(vim.eval('get(a:000, 0)'))):
                add_tt = True
            tokens = lex(cmdline, lexer)
            fmt = format(tokens, NvimFormatter(bg=vim.options['background']))
            return [
                ((lambda s, e, g, tt: (
                    (int(s), int(e), g)
                    + ((tt,) if add_tt else ())
                ))(*line.split()))
                for line in fmt.split('\n')[:-1]
            ]

    setup()


color_cmdline_setup()
```

```Vim
pyfile <sfile>:h/color_cmdline.py
function PygmentsFormat(cmdline, ...)
  return pyeval('color_cmdline_highlight()')
endfunction
let g:Nvim_color_cmdline = 'PygmentsFormat'
```

bfredl commented:
it might be worthwhile to think of how this will interact with #6162 later. I guess GUIs can recieve a list of pairs `[text, attrs]` with the attrs that apply to that chunk of text, alternatively just the text and the list of`[start, stop, attrs]`.

ZyX-I committed:
ex_getln: Add some more tests, fix some found errors

brcolow commented on test/functional/ui/cmdline_highlight_spec.lua:
```diff
@@ -55,47 +119,227 @@ describe('Command-line coloring', function()
+      :echo "«^                                |
+    ]])
+  end)
+  it('does the right thing when errorring', function()
```
typo: errorring -> erroring

ZyX-I committed:
eval,ex_getln: Add support for coloring input() prompts

ZyX-I commented:
Actually it appears that `input({opts})` is better then requiring prompt. What’s the point in requiring prompt if the official documentation states that you may have no prompt by supplying it an empty string, so basically prompt is not required?

bfredl commented:
I would guess the purpose is to nudge plugin authors to consider adding a prompt (instead of just calling `input()` with no args without even checking the docs that a prompt is possible), but with a dict they will need to check the docs anyway so then it makes sense to make it optional.

Also, having user commands and `input()` executing a callback on every input change is going to be useful for other purposes than coloring the cmdline, for instance live feedback in a window like `inccommand`. So I wonder if the callback argument really should be called `highlight` or `color`, maybe instead `callback` to acknowledge that it has other potential uses?

ZyX-I commented:
@bfredl One of the points of the coloring is “better have no colors then make user problems”. People using that for other purposes should be prepared that Neovim may reject calling their callback if it causes troubles: shows errors, takes too long to finish, alters some state, explicitly disabled by the user, etc. In the PR I am only going to disable it for errors, arabic shaping (not sure how to implement and how to test) and probably for some option set, but that may be extended depending on the problems I spot. If you know an existing facility to automatically interrupt after some time passed I would implement “too long to finish” as well.

bfredl commented:
> alters some state, explicitly disabled by the user, etc. In the PR I am only going to disable it for errors, arabic shaping (not sure how to implement and how to test) and probably for some option set, 

This would be fine for other use cases as long as one still can `rpcnotify()` (not `rpcrequest()`) and `start_timer(0, ...)` (currently the canonical way of "do something when the main loop is ready")

> If you know an existing facility to automatically interrupt

I don't know if `got_int` somehow can be abused for this. Checking `got_int` is quite common but calling `os_breakcheck` less so (which would be the natural place to check the time limit)

ZyX-I committed:
functests: Add missing wait()

ZyX-I committed:
ex_getln: Enable coloring for expression mode

ZyX-I committed:
functests: Abstract away some ways to enter cmdline coloring mode

Reason: should actually switch to using input() coloring because other coloring 
variants are eventually going away.

ZyX-I committed:
functests: Comment out failing test

ZyX-I committed:
functests: Make tests work with input()

There are still some issues: specifically, new “pending” test hangs busted.

ZyX-I committed:
functests: Fix typo

ZyX-I committed:
functests: Make “stops executing callback” test work

Needed to be adjusted to use input() (previously relied on side-effects of 
executing `:cmd`) and dismiss something (hidden “Press ENTER” message?).

ZyX-I committed:
ex_getln: Call highlight callback inside :try

ZyX-I committed:
functests: Partially uncomment <C-c> test

ZyX-I committed:
functests: Uncomment `{REDRAW}` part of “works” test

ZyX-I committed:
ex_getln: Fix “echoerr msg not shown” problem

This also attempted to fix problem with cancelling input() on error by avoiding 
standard error printing facilities (assumed thrown error message is the 
problem), but with no luck so far.

ZyX-I committed:
ex_getln: Make sure standard error reporting facility is not used

ZyX-I committed:
ex_getln: Fix indent

ZyX-I committed:
ex_getln: Do not make <C-c> interrupt input() after interrupting hl cb

ZyX-I committed:
ex_getln: Save and restore try state

Problem: when processing cycle such as

    :for pat in [' \ze*', ' \zs*']
    :  try
    :    let l = matchlist('x x', pat)
    :    $put ='E888 NOT detected for ' . pat
    :  catch
    :    $put ='E888 detected for ' . pat
    :  endtry
    :endfor

`:let l = …` throwing an error causes this error to be caught after 
color_cmdline attempts to get callback for highlighting next line (the one with 
`$put = 'E888 NOT…`). Saving/restoring state prevents this from happening.

bfredl commented on src/nvim/ex_getln.h:
```diff
@@ -52,6 +52,8 @@ typedef struct hist_entry {
   list_T *additional_elements;  ///< Additional entries from ShaDa file.
 } histentry_T;
 
+extern Callback getln_input_callback;
```
why not make it part of `save_ccline` ?

ZyX-I committed:
Merge branch 'master' into colored-cmdline

ZyX-I committed:
oldtests: Use $(TOOL) in place of $(VALGRIND)

ZyX-I committed:
api helpers: Save/restore more values in try_enter/try_leave

This fixes memory leak reported by ASAN. This also somehow fixes test40, though 
I have no idea why except that that test yields memory leak report.

ZyX-I committed:
ex_getln: Do not goto color_cmdline_end without first cleaning up

The issue with debug mode was actually not cleaning up after `try_enter`: 
location `&tstate` was pointing to got invalidated and received some “garbage” 
(actually, values that got stored on the stack afterwards). But pointer to that 
garbage was still stored in `msg_list`, so next attempt to check it resulted in 
a crash.

ZyX-I committed:
ex_getln: Do not do useless try_enter/try_leave calls

These are actually needed for two modes only. And even for these modes they 
should eventually go away.

ZyX-I committed:
doc: Adjust documentation a bit

2 spaces after sentense ends, clarify what will be highlighted.

ZyX-I committed:
ex_getln: Replace global with entry in save_ccline

ZyX-I committed:
api helpers: Also save and restore did_emsg

ZyX-I committed:
ex_getln: Check prev_prompt_errors before running redrawcmdline

Otherwise there will be infinite recursion and shortly a crash. Running 
redrawcmdline recursively occurs under color_cmdline_error label.

ZyX-I committed:
functests: Check for previously unchecked errors

ZyX-I committed:
functests: Test input() nesting support

ZyX-I committed:
*: Fix clint errors

ZyX-I committed:
ex_getln: Avoid GCC “unused variable” warning from QB

    17:25:45,363 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/ex_getln.c: In function ‘color_cmdline’:
    17:25:45,363 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/ex_getln.c:2335:8: error: variable ‘printed_errmsg’ set but not used [-Werror=unused-but-set-variable]
    17:25:45,363 WARN  - bool printed_errmsg = false;
    17:25:45,363 WARN  - ^
    17:25:45,399 WARN  - cc1: all warnings being treated as errors

ZyX-I committed:
functests: Add sleep to <C-c> test

ZyX-I commented:
Should be mostly ready. The only issue left I know about is making `<C-c>` interrupt test less flaky. Maybe just adding a sleep will be enough.

ZyX-I commented:
QB has the usual kbtree problem and that’s the only failure. AppVeyor got

```
[  FAILED  ] C:/projects/neovim/test/functional\core\job_spec.lua @ 267: jobs can pass user data to the callback
C:/projects/neovim/test/functional\core\job_spec.lua:271: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = {
      [l] = { ... more }
      [n] = 5
      [s] = 'str' }
   *[2] = {
      [1] = 'foo' } } }
Expected:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = {
      [l] = { ... more }
      [n] = 5
      [s] = 'str' }
   *[2] = {
      [1] = 'foo'
     *[2] = '' } } }
stack traceback:
	C:/projects/neovim/test/functional\core\job_spec.lua:271: in function <C:/projects/neovim/test/functional\core\job_spec.lua:267>
```

yet again, otherwise OK. Travis is still building, but so far everything succeeded and only coverage was left.

ZyX-I commented:
Travis succeeded.

bfredl commented on runtime/doc/eval.txt:
```diff
@@ -4723,6 +4724,23 @@ input({opts})
+		of the multibyte character (highlighting must not break 
+		multibyte characters), hl_end_byte + 1 may be equal to the 
+		input length.  Start column must be in range [0, len(input)), 
+		end column must be in range (hl_start_byte, len(input)], 
```
I don't think `(,]` notation is used elsewhere. Maybe "end byte must be greater than start byte in the same item". Also, always use "byte" and not mix "byte" with "column"?

ZyX-I replied:
This is ISO 31-11 standard mathematical notation for sets. Writing this in words would be rather lengthy and I pretty much sure I would not like the result.

bfredl replied:
but is it used elsewhere in the documentation? You write a similar constraint in words below and it doesn't seem to lengthy.

ZyX-I replied:
“Similar constraint” below is mere ≥. And it *is* lengthy: compare “≥” or “[ to “is greater then or equal to”, just not *too* lengthy and I also do not like writing “previous {varname}” in a math expression, it looks bad and not mathy.

bfredl replied:
My point is mainly of consistency: there is no need to use obscure math notation to save a few words in otherwise spelled out paragraph text (which is what I comment on, and not an equation)

ZyX-I replied:
It is a standard math notation, why do you say it is obscure? It saves lots of words, not “a few”. And it is already used in documentation: with incorrectly (does not find my own interval) constructed regex there is a match in `:h log()`. Actually, there is more then one occurrence in math functions’ descriptions.

bfredl commented on src/nvim/api/private/helpers.c:
```diff
@@ -37,6 +37,64 @@ typedef struct {
+  assert(!did_throw);
+  assert(!did_emsg);
+  assert(msg_list == &tstate->private_msg_list);
+  assert(*msg_list == NULL);
```
Is it intentional, if compiled without assertions, they will be silently ignored? (including leaking memory?)

ZyX-I replied:
Where do you see leaking memory? No assertions here are doing anything, but checking const comparisons results.

bfredl replied:
I mean if the disabled asserts would have failed, we will silently overwrite this state, including pointers that I presume points to allocated memory?

ZyX-I replied:
They are, but if asserts would’ve failed we are getting some kind of “undefined behaviour” and not “leaking memory”: failing assert indicates that code is being based on wrong assumptions. Fixing “memory leak” in this case will most likely just introduce another bug instead, like silently ignored exception.

bfredl replied:
That's what meant with "intentional", we can accept undefined exception state and small memory leak until the code elsewhere is fixed.

bfredl commented on test/functional/eval/input_spec.lua:
```diff
@@ -196,6 +228,19 @@ describe('input()', function()
+    feed([[:call input({"highlight": "RainBowParens"})<CR>]])
+    wait()
+    feed('(())')
+    wait()
```
wait shouldn't be used prior to `expect()` (I'm also suspicious of the one before, `screen:sleep()` is better if we really need to simulate "the user stopped typing for some while", but I wouldn't expect that here, `input()` should be sync in its operations)

ZyX-I replied:
Without the one before test fails.

bfredl replied:
do we know why? does `input()` flush the input buffer somehow?  Real interaction will not have this API call, so it would be good to identify the issue.

ZyX-I replied:
@bfredl I got that when I removed all the `wait()`s:

```
./test/functional/ui/screen.lua:302: Row 4 did not match.
Expected:
  |{EOB:~                        }|
  |{EOB:~                        }|
  |{EOB:~                        }|
  |*                         |
  |{RBP1:(}{RBP2:()}{RBP1:)}^                     |
Actual:
  |{EOB:~                        }|
  |{EOB:~                        }|
  |{EOB:~                        }|
  |*{RBP1:(}{RBP2:()}{RBP1:)}^                     |
  |                         |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.

stack traceback:
        ./test/functional/ui/screen.lua:302: in function 'wait'
        ./test/functional/ui/screen.lua:216: in function 'expect'
        test/functional/eval/input_spec.lua:233: in function <test/functional/eval/input_spec.lua:230>
```

I do not know why though. And I do not see any problems in TUI.

ZyX-I replied:
After adding a logger which checks how my function is being called I see that it is being called twice for each input. It is rather strange, this should indicate that `draw_cmdline` is called twice.

ZyX-I replied:
May be a good idea to cache previous highlighting result and reuse it if prompt_id and text to highlight matches what was used previously.

bfredl replied:
is it not just that the cmdline does not fit the screen width? If I make the terminal narrow I get the same behavior in TUI.

ZyX-I replied:
@bfredl In the test where I observed calling that twice I had 118 columns, no prompt and one to four characters (typed `(())` with a slightly adjusted RainBowParens function). In the input_spec while screen is rather narrow it still fits.

bfredl replied:
I was referring to the extra `wait()`,  which will not be needed if remove the narrow width issue (make the window wider, or use `command`)

I've also noticed that the cmdline redraw logic is not very efficient. We needed some throttling to implement the cursor type switching without overloading the ui protocol serialization with data....

ZyX-I replied:
I guess I understood: what is being too wide is the previous command, not the input. In the other (new) test file I omitted the issue by creating a mapping which calls `input`. I cannot use `command` though, it blocks.

bfredl replied:
you can use `nvim_async('command', ...)`

bfredl commented on src/nvim/ex_getln.c:
```diff
@@ -2374,8 +2641,23 @@ static void draw_cmdline(int start, int len)
-  } else
-    msg_outtrans_len(ccline.cmdbuff + start, len);
+  } else {
+draw_cmdline_no_arabicshape:
```
should we document that coloring is disabled with arabicshape? Or is it a limitiation we expect to remove soon?

ZyX-I committed:
doc: Update documentation

ZyX-I committed:
functests: Remove unneeded wait()s

oni-link commented on src/nvim/message.c:
```diff
@@ -1628,6 +1628,27 @@ void msg_puts_attr_len(const char *const str, const ptrdiff_t len, int attr)
+/// @param[in]  attr  Highlight attributes.
+/// @param[in]  fmt  Format string.
+void msg_printf_attr(const int attr, const char *const fmt, ...)
+  FUNC_ATTR_NONNULL_ALL
```
This attribute looks too restrictive. What if we want to print a pointer value `NULL`?

ZyX-I committed:
message: Only require second (format) argument to be not NULL

ZyX-I committed:
functests: Replace wait() with nvim_async

ZyX-I committed:
ex_getln: Cache highlight callback calling results

ZyX-I commented:
I have added caching of the highlight callback calling results. Pay attention to `-- (…)` in new tests: it appears that highlight callback (with command-line redrawing function) may be called *three* times in a row with one and the same cmdline.

ZyX-I committed:
Revert "functests: Replace wait() with nvim_async"

This reverts commit e129607988b88719935bc4af517e7ee2689f5871.

Tests stopped working in CI.

ZyX-I commented:
Travis hanged again.

ZyX-I committed:
Merge branch 'master' into colored-cmdline

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -126,7 +126,6 @@ Commands:
 Functions:
   |dictwatcheradd()| notifies a callback whenever a |Dict| is modified
   |dictwatcherdel()|
-  |execute()| works with |:redir|
```
why was this removed?

ZyX-I replied:
Vim has `execute()`, so it can’t be in vim_diff any longer. And in any case implementation has nothing to do with `:redir`.

justinmk replied:
The behavior of `execute(':redir')` differs. That's what this note is about.

ZyX-I replied:
Then what is it doing in “New features” section?

ZyX-I committed:
Merge branch 'master' into colored-cmdline

ZyX-I commented:
Travis passed.

bfredl commented on src/nvim/api/private/helpers.c:
```diff
@@ -37,6 +37,64 @@ typedef struct {
+  return ret;
+}
+
 /// Start block that may cause vimscript exceptions
```
Maybe clarify when this can be used as opposed to `try_enter`. IIUC command line could be colored in a situation where there could be a pending error (perhaps from a job/timer event handler active in cmdline mode?) so `try_enter` must be used, while an API function can only be called in a so far error free state, so `try_start` is safe?

ZyX-I replied:
Pending errors I saw so far were either from the command-line itself (there is a test with error in `:while` cycle body) or from things like mappings (there also is a test). If timers/etc may spawn errors like this then they should use `try_enter` as well.

bfredl commented on test/functional/eval/input_spec.lua:
```diff
@@ -196,6 +228,18 @@ describe('input()', function()
   end)
+  it('supports highlighting', function()
+    feed([[:call input({"highlight": "RainBowParens"})<CR>]])
+    wait()
```
this wait should still be removed. In any realistic situation, `input()` will be invoked with one call, not with an unrelated event incidentally right after it.

bfredl commented:
Otherwise LGTM

ZyX-I committed:
api/helpers: Clarify try_start() usage

ZyX-I committed:
functests: Remove wait() from input_spec

ZyX-I committed:
api/helpers: Fix typo

ZyX-I committed:
Merge branch 'master' into colored-cmdline

ZyX-I commented:
(Without the merge clint fails locally.)

ZyX-I commented:
AppVeyor failure looks unrelated:

```
[  FAILED  ] 1 test, listed below:
[  FAILED  ] C:/projects/neovim/test/functional\core\job_spec.lua @ 69: jobs changes to given `cwd` directory
C:/projects/neovim/test/functional\core\job_spec.lua:78: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = 0
   *[2] = {
      [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimdDir0n\0' } } }
Expected:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = 0
   *[2] = {
      [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimdDir0n\0'
     *[2] = '' } } }
stack traceback:
	C:/projects/neovim/test/functional\core\job_spec.lua:78: in function <C:/projects/neovim/test/functional\core\job_spec.lua:69>
```

Travis hanged in oldtests, test49:

```
!echo 'q^Mq' | $NVIM_PRG -u NONE -N -es -c 'debuggreedy|set viminfo+=nviminfo' -c 'let ExtraVimBegin = 1' -c 'let ExtraVimResult = "/tmp/nvimSdwGqV/9"' -c 'breakadd func 3 G' -S /tmp/nvimSdwGqV/8
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

Restarting.

justinmk commented:
My tentative observation is that the test49 hang seems to persist if travis cache is not cleared.

ZyX-I commented:
@justinmk This is planned to be merged after dealing with more urgent issues for the next (0.2.1) release?

justinmk commented:
@ZyX-I If you think it's ready, I think we should include it in 0.2.1

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -265,6 +273,8 @@ Lua interface (|if_lua.txt|):
 on cancel and completion respectively) via dictionary argument (replaces all 
 other arguments if used).
 
+|input()| and |inputdialog()| now support user-defined cmdline highlighting.
```
Why is only highlighting mentioned, in all of these docs? This is a handler that is called on each input character, correct? That's useful for much more than highlighting.

ZyX-I replied:
@bfredl asked this somewhere above: it is not guaranteed to be called, if real hooks cause any problems then I would rather add code which detects those problems and stops using hook then do anything. Currently this includes stopping calling the hook in case of arabic characters present. And handler is not called on each input character, it is called on *redraw*, minus cases when hook output can be taken from cache. I don’t have any information regarding what caused redraw, so one will have to compute a diff: hook always receives the whole drawn command-line. And whatever other uses you have in mind, note that hook is not supposed to edit input, `:echo` anything or issue any errors.

ZyX-I replied:
Though no, arabic characters fall under different category: “hard to implement, unlikely to happen and should really be solved on the different layer and not with some hacks in ex_getln.c and screen.c (it contains another incarnation of arabic shaping code)”. What falls into “causes problems” category now are hooks which throw errors.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -4703,6 +4703,7 @@ input({opts})
 		cancelreturn  ""       Same as {cancelreturn} from
 		                       |inputdialog()|. Also works with 
 		                       input().
+		highlight     nothing  Highlight handler: |Funcref|.
```
How often is the handler called? Should mention this somewhere.

Typically we prefix VimL handlers with `on_` (Vim does `_cb` suffix). Maybe this should be called `on_input`.

ZyX-I replied:
`on_input` is not right, it is currently called on redraw. And I do not like the hook or event view of the function, in my view it is more like a syntax definition, in a form of the VimL code: this allows things like caching in case of double highlight I implemented, additionally I don’t know command-line redrawing system enough to answer questions like “how many times and for which states highlight will be called when triggering `a` mapping defined like `cnoremap a bcd<C-\>e"efg"<CR>`” without experiments.

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -147,6 +146,15 @@ Highlight groups:
+  UI now supports command-line coloring. Officially only |input()| and 
+  |inputdialog()| may be colored, temporary for testing purposes expressions 
+  (e.g. |i_CTRL-R_=|) and regular command-line (|:|) are colored by callbacks 
+  defined in `g:Nvim_color_expr` and `g:Nvim_color_cmdline` respectively. 
```
`g:Nvim_` might not be a good convention because `sessionoptions+=globals` will try to store such variables if they are string/num types (doesn't apply here, but as a general convention ...).

Since these are temporary and deprecated, perhaps underscore prefix would make sense.

ZyX-I replied:
@justinmk I can’t have leading underscores, funcref variable names must start with a *capital*. Underscore is not one.

justinmk commented on test/functional/ui/cmdline_highlight_spec.lua:
```diff
@@ -0,0 +1,848 @@
+
+local screen
+
+-- Bug in input() handling: {REDRAW} will erase the whole prompt up until
```
should say `:redraw!` instead of `{REDRAW}` in this context.

ZyX-I committed:
functests: Check that input is correctly silenced

ZyX-I committed:
doc: Clarify how function is executed

ZyX-I committed:
functests: Alter comment

ZyX-I committed:
doc: State that it is called for new *displayed* input

justinmk commented on src/nvim/ex_getln.c:
```diff
@@ -2285,75 +2343,329 @@ void free_cmdline_buf(void)
+  } else if (colored_ccline->cmdfirstc == ':') {
+    try_enter(&tstate);
+    err_errmsg = N_(
+        "E5408: Unable to get Nvim_color_cmdline callback from g:: %s");
```
`g::` looks awkward. Suggestion:

    "E5408: Unable to get g:Nvim_color_cmdline callback: %s"

ZyX-I committed:
ex_getln: Make error messages look better

justinmk commented:
There's a non-trivial amount of logic involving `getln_interrupted_highlight`, `prompt_id`, related to concerns about performance. Just curious, why would performance be noticed here, typically commands occupy only 1 line, and AFAIK vim does not take such measures for insert-mode highlighting.

ZyX-I commented:
Insert-mode highlighting is done in C code. I am running VimL and expect it to be slow in cases like somebody trying to implement pygments in pure VimL. And I do not see too non-trivial amount: for `getln_interrupted_highlight` it is just “if `<C-c>` happened to arrive while highlighting assume that highlighting appeared to be too slow for the user to be disturbed, so no longer execute the highlight (`:try` catches interrupts as exceptions, so nothing is done specially here, they go with other errors), but do not assume that user meant to interrupt the prompt (what `getln_interrupted_highlight` is for)”. `prompt_id` is there not for performance, it disables running highlight in case of errors (in order to not spam user with errors), but makes the disablement not entirely permanent, only for one prompt.

ZyX-I commented:
Or, the other way around: `getln_interrupted_highlight` is what allows you to interrupt infinite cycle in highlight function, but *still* run what you want. Without it you could not do `let g:Nvim_color_cmdline = function('Halting')` and run any Ex mode command afterwards without hacks (there still is `<C-r>=` and `execute()`), including the one which will fix the problem.

`prompt_id` allows disabling highlighting for the duration of one prompt only, in the above case it is needed for you to not type `<C-c>` after typing each character of `unlet g:Nvim_color_cmdline`. Though I think I can remove `prompt_id` and just keep `prev_prompt_errors` and `last_ccline_colors` in `CmdlineInfo`. Behaviour will slightly change: in the example you would need to type second `<C-c>` if you use another prompt: `:u<C-c>nlet <C-r>="g"<CR>:<C-c>Nvim_color_cmdline<CR>`, and highlight function will not be recalled.

ZyX-I commented:
CI succeeded, only left running is gcov “flaky” build which will not possibly affect green status.

ZyX-I commented:
All “flaky” appeared to be green as well, it is uncommon.
